### DISCUSSION TITLE
「コマンド文」の定義と「コマンド文」単位の操作

### DISCUSSION THREAD

**Discussion by cmplstofB** [akinomyoga/ble.sh#90] [Ideas]
Emacsにおいて，「[文 (sentence)](https://www.gnu.org/software/emacs/manual/html_node/emacs/Sentences.html)」は「')', ']', '\'', '"'などを含む，行末にあるか後ろに二つの空白を持つ'.', '!', '?' で終わるもの」として定義されています。
Vimにおいて，「[文 (sentence)](https://vim-jp.org/vimdoc-ja/motion.html#sentence)」は「'.', '!', '?' で終わり、その後に行末かスペースもしくはタブが続くもの」として定義されています。また，このsentenceをテキストオブジェクトのように扱え，既定では[<kbd>(</kbd>](https://vim-jp.org/vimdoc-ja/motion.html#%28)・[<kbd>)</kbd>](https://vim-jp.org/vimdoc-ja/motion.html#%29)に割り当てられているモーションキーを用いて，<kbd>d</kbd><kbd>3</kbd><kbd>)</kbd>（後方3文を削除）のような編集が可能です。

一方，ble.shにおいて，というよりコマンドライン操作全般において，EmacsやVimが定義する終止符などを含んだsentenceは，VCSのコミットメッセージやヒアドキュメントなどの数少ない場面を除いてあらわれないだろうと予想します。ですので，EmacsやVimのsentenceそのものを正確に再現しても，あまり使い道がないように感じます。

ここで，sentenceの定義をコマンドライン独特のものに改変・拡張することを提案します。コマンドラインにおける「**コマンド文**」を
「
制御演算子および空白文字以外のメタ文字<sup class="footnote-ref"><a href="#fn1" id="fnref1">[1]</a></sup>で区切られた前記以外の字句の連続
」
と定義します。

コマンド文の例（`^`による連続する波線で示した部分がコマンド文です）
```console
$ echo $PATH | tr ':' '\n'
  ^^^^^^^^^^   ^^^^^^^^^^^
$ nohup sleep 100 > /dev/null &
  ^^^^^^^^^^^^^^^   ^^^^^^^^^
$ for v in "${BASH_VERSINFO[@]}"; do echo $v; done
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^  ^^^^
$ cat <<. | xargs -I @ echo @
foo
bar
baz
.
  ^^^       ^^^^^^^^^^^^^^^^^^
```

この「コマンド文」を，EmacsやVimが「文」に対して行うのと同じ操作を行えるようにする（例えばEmacs modeなら<kbd>M-a</kbd>で「コマンド文の先頭に移動する」）というのはどうでしょうか。
もしよろしければ，ご一考頂けるとうれしいです。

<ol>
<li id="fn1"><p>呼称はbashの日本語manページの定義・用語を流用しました。具体的には
`|`，`|&`，`&`，`;`，`(`，`)`，`<`，`>`，`||`，`&&`，`;;`です
<a href="#fnref1" class="footnote-backref">↩</a></p>
</ol>

**Comment by akinomyoga** [comment]
ご提案ありがとうございます!

```bash
$ nohup sleep 100 > /dev/null &
  ^^^^^^^^^^^^^^^   ^^^^^^^^^
```

redirection を独立したコマンドとして取り扱うという事は、文法的な構造まで立ち入らずに字句的に判定するという想定でしょうか。こちらでも今までに似たようなことを何となく思い浮かべたりすることもありましたが、もっと文法構造を反映したものを考えたりしていました。とはいいつつ実装が面倒そうだなあと思ったり。まあ、どちらの実装にしても quote された `|` や `&` 等を除外しないと直感的な動作にならないと思うので、文法情報を参照しなければならないことには変わりなさそうです。

ここでは実装の面倒さだとかは考えない事にして自然な実装についてもう少し考えてみたいですね。例えば、以下のような場合にどの様に振る舞うと良さそうですか?

```bash
cmd1 $(cmd2 A | cmd3 B) C
```

`A` の位置で <kbd>M-a</kbd> (先頭に移動) をすると `cmd2` の先頭に、`B` の位置で <kbd>M-a</kbd> すると `cmd3` の先頭に行くというところは異論はまあないと思います。問題は `C` にいる時に <kbd>M-a</kbd> をした時に `)` の直後に移動するのが良いか、或いは `cmd1` の先頭に移動するのが良いのかという点でしょうか。

論理的に考えたら `cmd1` の先頭に移動するのが自然な気がしますけれど、移動の利便性という観点から考えるともっと細かく移動できた方が良いかもしれませんね。

**Reply by cmplstofB** [reply]
（返信遅れてすいません）

> 文法的な構造まで立ち入らずに字句的に判定する

引用していただいた例については，単にメタ文字の定義に従って愚直に示しただけで，「コマンド文」の解析を字句的に行うか構文的に行うかはとくに考えていませんでした。

一方，全体的な考えとして，案元の〝（EmacsやVimにおける）普通の文〟は意味的にではなく字句的に定義されているので，それに従う，というのが自然な気もします。
とはいえ，コマンド行とは違って普通の文を意味的に定義しようとすると自然言語処理が必要になるので，EmacsやVimでは技術的な面から字句的な実装にしているだけかもしれませんが。


> 以下のような場合

[複合コマンド](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_04)を生成するような構文，例えば`$()`や`{}`などは，普通の文における引用符の類推とし，案元の実装における引用符を跨ぐ文と同じような挙動にする，というのはどうでしょうか。

```
His statement "I am telling the truth."  was a lie.  And it was crucial.
                                         ^ᵇ    ^ᵃ
```

上例のような文をEmacsで編集する際，波線aの位置で<kbd>M-a</kbd>とやると波線bの位置に移動します。
ですので

```console
cmd1 $(cmd2 A | cmd3 B) C
                      ^ᵇ^ᵃ
```

波線aで「コマンド文の先頭に移動」した場合は波線bの位置に移動するのがよいかなと思います。
コマンド文は通常の文とは定義こそ違いますが，利用者にとっては「（普段使っているeditorの）文として扱われる」ような挙動が最も適切ではないでしょうか。

**Reply by akinomyoga** [reply]
お返事ありがとうございます。自分から質問しておいて何ですが、ちょっと考えてみたところ `)` の直前で止めるようにするというのは well-defined かどうか分からなくなってきました…。後、実装的にも、(文法構造に忠実に移動するよりも) 実は却って面倒かもしれません。ただ、具体的にどのような物を想定されているかにもよります。例えば以下のような場合にどのような動作になるのが良いでしょうか:

```bash
cmd1 "abc$(cmd2 arg1)xyz" arg2 CurPos
cmd1 "abc${v:-$(cmd2 arg1)}xyz" arg2 CurPos
cmd1 'abc(hello)xyz' arg2 CurPos
cmd1 'abc$(hello)xyz' arg2 CurPos
a[$(cmd2)]=1 cmd1 CurPos
```

但し、`CurPos` の位置にカーソルを於いて其処で <kbd>M-a</kbd> (beginning-of-command) を実行するという想定です。
